### PR TITLE
BUG: break reference cycle in Index._engine

### LINKED ISSUE #27585: Index._engine creates cyclic reference
#### Code Sample, a copy-pastable example if possible

```python
import gc
import pandas as pd

gc.disable()

idx = pd.Int64Index(data=[0])

# trigger a call to idx._engine using public API, but a unit test might also be
# able to use the private API here
idx.is_monotonic_increasing
del idx

remaining_indices = [o for o in gc.get_objects() if isinstance(o, pd.Int64Index)]
assert remaining_indices == []
```
#### Problem description

The (indirect) call to `idx._engine` creates a cyclic reference, therefore indices are not removed without calling the GC:

```python
import objgraph
objgraph.show_backrefs(
    remaining_indices,
    filename='obj.png',
    shortnames=False,
    max_depth=10,
)
```

obj

On certain payloads (e.g. on `dask.distributed` clusters), this might increase memory consumptions significantly.

#### Expected Output
Index is cleared on `del idx` call.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.6.6.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.9.125-linuxkit
machine          : x86_64
processor        :
byteorder        : little
LC_ALL           : en_US.UTF-8
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 0.25.0
numpy            : 1.16.4
pytz             : 2019.1
dateutil         : 2.8.0
pip              : 10.0.1
setuptools       : 39.1.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None

</details>

### PR DESCRIPTION
**Pull request by crepererum**
- [X] closes #27585
- [X] tests added / passed
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
this is much better done by removing ``@cache_readonly`` and then use a weakref with the same types of caching (because cache_readonly does this in a very specific way)

**Comment by crepererum** [comment]
@jreback I don't really understand what you mean. The issue is not the cache but that `self._engine_type` (e.g. `Int64Engine`) holds a callable that refers `self` (i.e. the `Index` instance). Putting the result of `self._engine_type(...)` into a weakref will clear it to quickly and won't cache it. I could use a callable for `self._engine_type` that uses a weakref to `self` instead of the `partial` solution that I am using now, but I don't see how this relates to `@cache_readonly`. Can you please be more specific about your suggestion? Thanks in advance.

**Comment by jreback** [comment]
> @jreback I don't really understand what you mean. The issue is not the cache but that `self._engine_type` (e.g. `Int64Engine`) holds a callable that refers `self` (i.e. the `Index` instance). Putting the result of `self._engine_type(...)` into a weakref will clear it to quickly and won't cache it. I could use a callable for `self._engine_type` that uses a weakref to `self` instead of the `partial` solution that I am using now, but I don't see how this relates to `@cache_readonly`. Can you please be more specific about your suggestion? Thanks in advance.

and I am not really sure what you suggested actually does anything. so let's revisit. what is the actual problem? having a cyclic reference in not a bug.

**Comment by crepererum** [comment]
> and I am not really sure what you suggested actually does anything. so let's revisit. what is the actual problem? having a cyclic reference in not a bug.

You're right. I should have labeled it as a performance optimization. The issue with the reference cycle is that (similar to the one we had some time ago with the cycle between the indexer and the dataframe) that indices are not cleared from memory w/o running the GC. Under notebook / low-load conditions, this might not be an issue. For high-load system like dask / dask.distributed, this can be an issue since the python interpreter requires more memory than it should (peak memory consumption). I have actually observed exactly this issue under dask.distributed and depending on the load pattern, users can expect a temporary (i.e. until the GC finds the Index) over a gigabyte per worker. So what I would like to archive with this PR is that the index does not have a reference cycle and is cleared instantly when not used anymore.

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -687,7 +695,7 @@ def _cleanup(self):
     @cache_readonly
     def _engine(self):
         # property, for now, slow to look up
-        return self._engine_type(lambda: self._ndarray_values, len(self))
+        return self._engine_type(partial(_return_obj, self._ndarray_values), len(self))
```
Comment (jbrockmendel) [inline pandas/core/indexes/base.py]: would this work if return_obj were defined within `_engine`?

Is return_obj going to be useful in other places too?

**Comment by jreback** [comment]
> > and I am not really sure what you suggested actually does anything. so let's revisit. what is the actual problem? having a cyclic reference in not a bug.
> 
> You're right. I should have labeled it as a performance optimization. The issue with the reference cycle is that (similar to the one we had some time ago with the cycle between the indexer and the dataframe) that indices are not cleared from memory w/o running the GC. Under notebook / low-load conditions, this might not be an issue. For high-load system like dask / dask.distributed, this can be an issue since the python interpreter requires more memory than it should (peak memory consumption). I have actually observed exactly this issue under dask.distributed and depending on the load pattern, users can expect a temporary (i.e. until the GC finds the Index) over a gigabyte per worker. So what I would like to archive with this PR is that the index does not have a reference cycle and is cleared instantly when not used anymore.
> 
> If we agree on that change, I'll change the changelog entry and commit message before merging to make sure this is not labeled as a bug but as an enhancement.

not objecting to the change in principle. But not sure how adding an indirection different from the lambda helps here.

File: pandas/core/indexes/base.py (hunk 1 above)
Comment (crepererum) [reply pandas/core/indexes/base.py]: You're right, the nested function / lambda is NOT the issue, but the problem is that `self` is bound to the `lambda`. I've added a more elegant solution.

**Comment by crepererum** [comment]
> But not sure how adding an indirection different from the lambda helps here.

I've changed it to a more elegant solution and added a comment why this is helpful.

File: doc/source/whatsnew/v0.25.1.rst [hunk 2]
```diff
@@ -85,6 +85,7 @@ Interval
 Indexing
 ^^^^^^^^
 
+- Break reference cycle in :meth:`Index._engine` (:issue:`27585`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.25.1.rst]: Index._engine isn't part of the API docs. Phrase as "Break reference cycle involving :class:\`Index\`."

File: doc/source/whatsnew/v0.25.1.rst [hunk 3]
```diff
@@ -85,6 +85,7 @@ Interval
 Indexing
 ^^^^^^^^
 
+- Break reference cycle involving :class:`Index`. (:issue:`27585`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.1.rst]: can you elaborate a bit more on what this does (allow non-cycle gc)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
can you add an asv for this as well (the memory kind)

**Comment by crepererum** [comment]
> can you add an asv for this as well (the memory kind)

Did so. Not sure if this is the right place, but it demonstrates the issue as well:

Before (current master):
```
$ asv run -e -E existing --bench gc
· Discovering benchmarks
· Running 1 total benchmarks (1 commits * 1 environments * 1 benchmarks)
[  0.00%] ·· Benchmarking existing-py_...
[100.00%] ··· index_object.GC.peakmem_gc_instances
[100.00%] ··· ======== ======
               param1
              -------- ------
                 1      125M
                 2      131M
                 5      155M
              ======== ======
```

After (this PR):
```
$ asv run -e -E existing --bench gc
· Discovering benchmarks
· Running 1 total benchmarks (1 commits * 1 environments * 1 benchmarks)
[  0.00%] ·· Benchmarking existing-py_...
[100.00%] ··· index_object.GC.peakmem_gc_instances
[100.00%] ··· ======== ======
               param1
              -------- ------
                 1      125M
                 2      125M
                 5      125M
              ======== ======
```

File: doc/source/whatsnew/v0.25.1.rst [hunk 4]
```diff
@@ -85,6 +85,7 @@ Interval
 Indexing
 ^^^^^^^^
 
+- Break reference cycle involving :class:`Index` to allow gargabe collection of :class:`Index` objects without running the GC. (:issue:`27585`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.1.rst]: typo here

**Review by jreback — APPROVED** [review]
tiny comment, otherwise lgtm. don't push yet though, the CI is acting up.

**Comment by jreback** [comment]
lgtm. you have some lint issues I think. ping on green.

**Comment by crepererum** [comment]
@jreback ping :)

**Comment by TomAugspurger** [comment]
Thanks @crepererum!

**Pull request merged**